pkg/controller/volume/expand/cache/volume_resize_map.go

```diff
@@ -0,0 +1,285 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/golang/glog"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/kubernetes/pkg/util/strings"
+	"k8s.io/kubernetes/pkg/volume"
+	"k8s.io/kubernetes/pkg/volume/util/types"
+)
+
+// VolumeResizeMap defines an interface that serves as a cache for holding pending resizing requests
+type VolumeResizeMap interface {
+	AddPvcUpdate(newPvc *v1.PersistentVolumeClaim, oldPvc *v1.PersistentVolumeClaim, spec *volume.Spec)
+	GetPvcsWithResizeRequest() []*PvcWithResizeRequest
+	// Mark this volume as resize
+	MarkAsResized(*PvcWithResizeRequest) error
+	MarkForFileSystemResize(*PvcWithResizeRequest) error
+	MarkResizeFailed(*PvcWithResizeRequest, string) error
+}
+
+type volumeResizeMap struct {
+	// map of unique pvc name and resize requests that are pending or inflight
+	pvcrs map[types.UniquePvcName]*PvcWithResizeRequest
+	// kube client for making API calls
+	kubeClient clientset.Interface
+	// for guarding access to pvcrs map
+	sync.RWMutex
+}
+
+// PvcWithResizeRequest struct defines data structure that stores state needed for
+// performing file system resize
+type PvcWithResizeRequest struct {
+	// PVC that needs to be resized
+	PVC *v1.PersistentVolumeClaim
+	// Volume spec object that can be used from volume plugins
+	VolumeSpec *volume.Spec
```

Comment:
Why VolumeSpec? Do you plan to resize inline volumes too?

Reply:
This is not inline volume spec. This volume spec is one defined in volume package which plugins know how to consume. Typically, most plugin calls require `volume.Spec` starting with `FindVolumePluginBySpec` .

Reply:
To me it would look better if you store PV here (because that's what you have from informers) and create VolumeSpec just before calling volume plugin.

Reply:
done

Reply:
I still see `VolumeSpec *volume.Spec`

Reply:
hmm, why though?